src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.cs

```diff
@@ -174,49 +176,92 @@ private static Task<HttpResponseMessage> InnerSendAsync(HttpRequestMessage reque
+                            response.Dispose();
+                            response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
+
+                            // Retry in case of nonce timeout in server.
+                            if (TryGetRepeatedChallenge(response, challenge.SchemeName, isProxyAuth, out string challengeData))
                             {
-                                response.Dispose();
-                                response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
+                                digestResponse = new DigestResponse(challengeData);
+                                if (IsServerNonceStale(digestResponse) &&
+                                    await TrySetDigestAuthToken(request, challenge.Credential, digestResponse, isProxyAuth).ConfigureAwait(false))
+                                {
+                                    response.Dispose();
+                                    response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
+                                }
                             }
                         }
-                    }
-                }
-                else if (challenge.AuthenticationType == AuthenticationType.Basic)
-                {
-                    if (!preAuthenticate)
-                    {
-                        SetBasicAuthToken(request, challenge.Credential, isProxyAuth);
+                        break;
+
+                    case AuthenticationType.Basic:
+                        if (performedBasicPreauth)
+                        {
+                            break;
+                        }
 
                         response.Dispose();
+                        SetBasicAuthToken(request, challenge.Credential, isProxyAuth);
                         response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
-                    }
+
+                        if (preAuthenticate)
+                        {
+                            switch (response.StatusCode)
+                            {
+                                case HttpStatusCode.ProxyAuthenticationRequired:
+                                case HttpStatusCode.Unauthorized:
+                                    break;
+
+                                default:
+                                    lock (pool.PreAuthCredentials) // TODO #28045: Get rid of this lock.
+                                    {
+                                        try
+                                        {
+                                            pool.PreAuthCredentials.Add(authUri, BasicScheme, challenge.Credential);
```

Comment:
Memory leak, you never remove entries. HWR had a similar bug where it would add an entry for /a, /a/b, /a/b/c, etc.. We changed it to a LRU list with a max of 100 entries: https://referencesource.microsoft.com/#System/net/System/Net/_PrefixLookup.cs,1464f3b5a82af37f

Reply:
> you never remove entries

The other .NET Core handlers all use CredentialCache, so this is no worse than those.  As noted in https://github.com/dotnet/corefx/issues/28045, we should ideally have a different type anyway that doesn't require taking a giant lock; we can address any such concerns as part of that.